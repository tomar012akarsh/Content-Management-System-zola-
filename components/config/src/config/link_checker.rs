use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum LinkCheckerLevel {
    #[serde(rename = "error")]
    Error,
    #[serde(rename = "warn")]
    Warn,
}

impl Default for LinkCheckerLevel {
    fn default() -> Self {
        Self::Error
    }
}

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default)]
pub struct LinkChecker {
    /// Skip link checking for these URL prefixes
    pub skip_prefixes: Vec<String>,
    /// Skip anchor checking for these URL prefixes
    pub skip_anchor_prefixes: Vec<String>,
    /// Emit either "error" or "warn" for broken internal links (including anchor links).
    pub internal_level: LinkCheckerLevel,
    /// Emit either "error" or "warn" for broken external links (including anchor links).
    pub external_level: LinkCheckerLevel,
}
