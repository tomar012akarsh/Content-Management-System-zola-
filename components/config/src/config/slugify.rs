use serde::{Deserialize, Serialize};

use utils::slugs::SlugifyStrategy;

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default)]
pub struct Slugify {
    pub paths: SlugifyStrategy,
    pub paths_keep_dates: bool,
    pub taxonomies: SlugifyStrategy,
    pub anchors: SlugifyStrategy,
}
