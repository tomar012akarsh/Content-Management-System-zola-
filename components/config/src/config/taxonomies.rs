use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default)]
pub struct TaxonomyConfig {
    /// The name used in the URL, usually the plural
    pub name: String,
    /// The slug according to the config slugification strategy
    pub slug: String,
    /// If this is set, the list of individual taxonomy term page will be paginated
    /// by this much
    pub paginate_by: Option<usize>,
    pub paginate_path: Option<String>,
    /// Whether the taxonomy will be rendered, defaults to `true`
    pub render: bool,
    /// Whether to generate a feed only for each taxonomy term, defaults to `false`
    pub feed: bool,
}

impl Default for TaxonomyConfig {
    fn default() -> Self {
        Self {
            name: String::new(),
            slug: String::new(),
            paginate_by: None,
            paginate_path: None,
            render: true,
            feed: false,
        }
    }
}

impl TaxonomyConfig {
    pub fn is_paginated(&self) -> bool {
        if let Some(paginate_by) = self.paginate_by {
            paginate_by > 0
        } else {
            false
        }
    }

    pub fn paginate_path(&self) -> &str {
        if let Some(ref path) = self.paginate_path {
            path
        } else {
            "page"
        }
    }
}
