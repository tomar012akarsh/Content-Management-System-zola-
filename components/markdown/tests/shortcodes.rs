use config::Config;

mod common;

#[test]
fn can_render_simple_text_with_shortcodes() {
    let body = common::render(
        r#"
hello {{ out_put_id(id="shortcode-id") }}

{% quote() %}
A quote
{% end %}

{{ out_put_id(id="shortcode-id2") }}

{{ out_put_id(id="shortcode-id3") }}
    "#,
    )
    .unwrap()
    .body;
    insta::assert_snapshot!(body);
}

#[test]
fn can_grab_lang_in_html_shortcodes() {
    let body = common::render(
        r#"
hello in {{ i18n() }}
    "#,
    )
    .unwrap()
    .body;
    insta::assert_snapshot!(body);
}

#[test]
fn can_grab_lang_in_md_shortcodes() {
    let body = common::render(
        r#"
{{ book() }}
    "#,
    )
    .unwrap()
    .body;
    insta::assert_snapshot!(body);
}

#[test]
fn can_render_body_shortcode_and_paragraph_after() {
    let body = common::render(
        r#"
{% quote() %}
This is a quote
{% end %}

Here is another paragraph.
    "#,
    )
    .unwrap()
    .body;
    insta::assert_snapshot!(body);
}

#[test]
fn can_render_two_body_shortcode_and_paragraph_after_with_line_break_between() {
    let body = common::render(
        r#"
{% quote() %}
This is a quote
{% end %}

{% quote() %}
This is a quote
{% end %}

Here is another paragraph.
    "#,
    )
    .unwrap()
    .body;
    insta::assert_snapshot!(body);
}

#[test]
fn doesnt_render_ignored_shortcodes() {
    let body = common::render(
        r#"
{{/* youtube(id="w7Ft2ymGmfc") */}}
    "#,
    )
    .unwrap()
    .body;
    insta::assert_snapshot!(body);
}

// https://github.com/Keats/gutenberg/issues/522
#[test]
fn doesnt_try_to_highlight_content_from_shortcode() {
    let body = common::render(
        r#"
{{ four_spaces() }}
    "#,
    )
    .unwrap()
    .body;
    insta::assert_snapshot!(body);
}

#[test]
fn can_emit_newlines_and_whitespace_with_shortcode() {
    let body = common::render(
        r#"
{% pre() %}
Hello

Zola

!

{% end %}
    "#,
    )
    .unwrap()
    .body;
    insta::assert_snapshot!(body);
}

#[test]
fn can_passthrough_markdown_from_shortcode() {
    let body = common::render(
        r#"
Hello

{% md_passthrough() %}
# Passing through

*to* **the** document
{% end %}
    "#,
    )
    .unwrap()
    .body;
    insta::assert_snapshot!(body);
}

// https://github.com/getzola/zola/issues/1172
#[test]
fn doesnt_escape_html_shortcodes() {
    let body = common::render(
        r#"
{{ image(alt="something") }}
    "#,
    )
    .unwrap()
    .body;
    insta::assert_snapshot!(body);
}

#[test]
fn errors_on_unknown_shortcodes() {
    let body = common::render(
        r#"
{{ unknown() }}
    "#,
    );
    assert!(body.is_err());
}

// https://github.com/getzola/zola/issues/1172
#[test]
fn can_render_commented_out_shortcodes() {
    let body = common::render(
        r#"
<!-- {{ image(alt="something") }} -->
    "#,
    )
    .unwrap()
    .body;
    insta::assert_snapshot!(body);
}

#[test]
fn invocation_count_increments_in_shortcode() {
    let body = common::render(
        r#"
{{ a() }}
{{ b() }}
{{ a() }}
{{ b() }}
    "#,
    )
    .unwrap()
    .body;
    insta::assert_snapshot!(body);
}

// https://github.com/getzola/zola/issues/1689
#[test]
fn html_shortcode_regression() {
    let inputs = vec![
        r#"{{ ex2(page="") }} {{ ex1(page="") }} {{ ex3(page="std") }}"#,
        r#"<p>{{ ex2(page="") }} {{ ex1(page="") }} {{ ex3(page="std") }}</p>"#, // in html
        r#"<p>\n{{ ex2(page='') }}\n</p>"#,                                      // with newlines
        r#"<span>{{ ex2(page='') }}</span>\n**The Book** {{ ex2(page='') }}"#,
        r#"a.{{ ex2(page="") }} b.{{ ex1(page="") }} c.{{ ex3(page="std") }}"#,
    ];

    for input in inputs {
        let body = common::render(input).unwrap().body;
        insta::assert_snapshot!(body);
    }
}

#[test]
fn can_split_shortcode_body_lines() {
    let body = common::render(
        r#"
{% split_lines() %}
multi
ple
lines
{% end %}
    "#,
    )
    .unwrap()
    .body;
    insta::assert_snapshot!(body);
}

#[test]
fn can_render_shortcodes_with_tabs() {
    // This can cause problems mostly because the 4 spaces sometimes used for tabs also are used
    // to indicate code-blocks
    let body = common::render(
        r#"
{{ with_tabs() }} {{ with_tabs() }}
    "#,
    )
    .unwrap()
    .body;
    insta::assert_snapshot!(body);
}

// https://github.com/getzola/zola/issues/1355
#[test]
fn can_render_list_with_shortcode() {
    let body = common::render(
        r#"
* a
* b
    {{ with_tabs() }}
* c
    {{ with_tabs() }}
    "#,
    )
    .unwrap()
    .body;
    insta::assert_snapshot!(body);
}

// https://github.com/getzola/zola/issues/1655
#[test]
fn shortcodes_do_not_generate_paragraphs() {
    let body = common::render(
        r#"
{% web_component() %}
some code;
more code;

other code here;
{% end %}
    "#,
    )
    .unwrap()
    .body;
    insta::assert_snapshot!(body);
}

#[test]
fn can_render_markdown_in_shortcodes() {
    let mut config = Config::default_for_test();
    config.markdown.highlight_code = true;
    let body = common::render_with_config(
        r#"
{% render_md() %}

```
some code;
```

{% end %}
    "#,
        config,
    )
    .unwrap()
    .body;
    insta::assert_snapshot!(body);
}

// https://github.com/getzola/zola/issues/1600
#[test]
fn can_use_shortcodes_in_quotes() {
    let body = common::render(
        r#"
> test quote
> {{ image(alt="a quote") }}
    "#,
    )
    .unwrap()
    .body;
    insta::assert_snapshot!(body);
}
