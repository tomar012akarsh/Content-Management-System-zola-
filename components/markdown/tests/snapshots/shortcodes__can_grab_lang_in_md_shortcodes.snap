---
source: components/markdown/tests/shortcodes.rs
expression: body
---
<p><img src="https://www.getzola.org/test/cover.en.png" alt="Book cover in en" /></p>

